### PR TITLE
Qscript provenance

### PR DESCRIPTION
**Pull request by alissapajer**

### CODE REVIEW

**Comment by alissapajer** [comment]
@sellout here's the current state of things, rebased on master. There are a couple minor FIXME's that we could land with this PR as well.

File: core/src/main/scala/quasar/qscript/QScript.scala [hunk 1]
```diff
@@ -307,318 +372,360 @@ class Transform[T[_[_]]: Recursive: Corecursive: EqualT: ShowT, F[_]: Traverse](
...
-          val orderStrs: PlannerError \/ List[String] = rebase(rebasedArrays, order) match {
-            case ConcatArraysN(as) => as.traverse(StrLit.unapply(_)) \/> InternalError("unsupported ordering type") // disjunctionify
-            case StrLit(str) => List(str).right
+      val directionsList: PlannerError \/ List[SortDir] = {
+        val orderStrs: PlannerError \/ List[String] =
+          directions.ana(CoEnv.freeIso[Unit, MapFunc[T, ?]].reverseGet).project match {
+            case ConcatArraysN(as) => as.traverse { a =>
+              StrLit.unapply(a.cata(CoEnv.freeIso[Unit, MapFunc[T, ?]].get))
+            } \/> InternalError("unsupported ordering type")
+            //case StrLit(str) => List(str).right // FIXME catch this case too
```
Comment (alissapajer) [inline core/src/main/scala/quasar/qscript/QScript.scala]: @sellout I'd like this FIXME to land with this PR - if you know the quick fix, go for it, else we can pair on it later!

File: core/src/main/scala/quasar/qscript/QScript.scala [hunk 2]
```diff
@@ -307,318 +372,349 @@ class Transform[T[_[_]]: Recursive: Corecursive: EqualT: ShowT, F[_]: Traverse](
...
+          rebaseBranch(left, lMap).mapSuspension(FI.compose(envtLowerNT)),
+          rebaseBranch(right, rMap).mapSuspension(FI.compose(envtLowerNT)),
+          equiJF,
+          Inner,
+          Free.point(LeftSide))))).right
+
+    case LogicalPlan.InvokeFUnapply(set.Except, Sized(a1, a2)) =>
+      val (src, lMap, rMap, left, right) = merge(a1, a2)
+      EnvT((
+        src.project.ask, // TODO is this the correct provenance?
```
Comment (alissapajer) [inline core/src/main/scala/quasar/qscript/QScript.scala]: @sellout I wanted to ask for your thoughts on this one. `Except` is not affecting provenance - is that what you'd expect?

File: core/src/main/scala/quasar/qscript/QScriptCore.scala [hunk 3]
```diff
@@ -137,65 +138,59 @@ object QScriptCore {
...
+            // val (newRep, lrep, rrep) = concat(rep1, rep2.map(_ + func1.length))
+            val mapL = bucket1 >> left
+            val mapR = bucket2 >> right
+
+            (mapL ≟ mapR).option(
+              SrcMerge[EnvT[Ann[T], QScriptCore[IT, ?], Unit], FreeMap[IT]](
+                EnvT((Ann(b1, UnitF),
+                  Reduce((),
+                    mapL,
+                    // FIXME: Concat these things!
```
Comment (alissapajer) [inline core/src/main/scala/quasar/qscript/QScriptCore.scala]: This is an old one that got forgotten about.

File: core/src/main/scala/quasar/qscript/QScriptCore.scala [hunk 4]
```diff
@@ -137,65 +138,59 @@ object QScriptCore {
...
+            // val (newRep, lrep, rrep) = concat(rep1, rep2.map(_ + func1.length))
+            val mapL = bucket1 >> left
+            val mapR = bucket2 >> right
+
+            (mapL ≟ mapR).option(
+              SrcMerge[EnvT[Ann[T], QScriptCore[IT, ?], Hole], FreeMap[IT]](
+                EnvT((Ann(b1, HoleF),
+                  Reduce(SrcHole,
+                    mapL,
+                    // FIXME: Concat these things!
```
Comment (alissapajer) [inline core/src/main/scala/quasar/qscript/QScriptCore.scala]: ok, here's the place where we don't have concat on `Sized` / `Fin` working

File: core/src/main/scala/quasar/qscript/QScript.scala (hunk 2 above)
Comment (sellout) [inline core/src/main/scala/quasar/qscript/QScript.scala]: It shouldn’t affect provenance, but it needs to track the provenance of the left side. This is currently tracking the provenance of the common src.

File: core/src/main/scala/quasar/qscript/QScriptCore.scala (hunk 4 above)
Comment (sellout) [inline core/src/main/scala/quasar/qscript/QScriptCore.scala]: Well, this’ll be easier without shapeless ;)

**Pull request closed without merging**